dspy/teleprompt/finetune_teleprompter.py

```diff
@@ -0,0 +1,377 @@
+
+import dspy
+from dspy.utils.logging import logger
+from dspy.evaluate.evaluate import Evaluate
+from dspy.primitives.example import Example
+from dspy.primitives.program import Program
+from dspy.primitives.prediction import Prediction
+
+
+#-------------------------------------------------------------------------------
+#    Templates for the user-facing strings used by this module
+#-------------------------------------------------------------------------------
+
+_INFO_DEFAULT_TEACHER = """No teacher provided. Using a copy of the student \
+program as the teacher."""
+
+_INFO_STRUCTURAL_EQUIVALENCY = """Ensuring that the student and teacher are \
+are structurally equivalent."""
+
+_INFO_SHARED_PREDICTOR = """Ensuring that the student and teacher programs do \
+not share predictors."""
+
+_INFO_LM_CONSISTENCY = """Ensuring that the teacher program satisfies the LM \
+consistency property."""
+
+_INFO_BOOTSTRAP_DATA = """Bootstrapping data on {} examples with the program \
+{}, with {} threads"""
+
+#-------------------------------------------------------------------------------
+#    Helper functions
+#-------------------------------------------------------------------------------
+
+# TODO: Shared below are useful functions. Similar procedures are implemented
+# separately and used by other DSPy teleprompters. These can be moved to shared
+# locations.
+def prepare_teacher(
+        student: Program,
+        teacher: Program = None
+    ) -> Union[Program, AssertionError]:
+    """Prepare the teacher program with respect to the student program.
+    
+    Args:
+        student: The student program.
+        teacher: The teacher program. If `None`, a copy of the student program
+            is used as the teacher. Defaults to `None`.
+
+    Returns:
+        The copied teacher program.
+    
+    Raises:
+        AssertionError: If the teacher is not an instance of the Program class.
+    """
+    # If teacher is None, use a copy of the student program as the teacher
+    if teacher is None:
+        logger.info(_INFO_DEFAULT_TEACHER)
+        teacher = student.deepcopy()
+    else:
+        teacher = teacher.deepcopy()
+
+    # Ensure that the student and teacher programs have the same structure
+    logger.info(_INFO_STRUCTURAL_EQUIVALENCY)
+    student._assert_structural_equivalency(teacher)
+
+    # Ensure that the predictors of the programs point to different objects
+    logger.info(_INFO_SHARED_PREDICTOR)
+    student._assert_no_shared_predictor(teacher)
+
+    # Ensure that the LM consistency property is satisfied
+    logger.info(_INFO_LM_CONSISTENCY)
+    teacher._assert_lm_consistency()
+
+    # If the global LM is being used, set it to the LMs of the copied teacher
+    # program predictors to to avoid handling the same edge cases later
+    if dspy.settings.lm:
+        teacher._set_all_predictor_lms(dspy.settings.lm)
+
+    return teacher
+
+
+# TODO: fix docstring
```

Comment:
Having no docstring is better than having a bad docstring. The library is still evolving, I'd vote for not writing long docstrings in general until super stable.

Reply:
This is especially true for internal methods like this, which are not consumed by our users.

Reply:
Removed -- originally meant for public use